File: DSA/Stack/Code/PostfixToPrefix.cs
// Postfix to Prefix in C#

using System;
using System.Collections.Generic;

class PostfixToPrefix {
    static bool IsOperator(char c) {
        return (c == '+' || c == '-' || c == '*' || c == '/');
    }
    
    static string PostfixToPrefixConversion(string postfix) {
        Stack<string> stack = new Stack<string>();
        
        foreach (char c in postfix) {
            if (IsOperator(c)) {
                string op2 = stack.Pop();
                string op1 = stack.Pop();
                
                string temp = c + op1 + op2;
                stack.Push(temp);
            } else {
                stack.Push(c.ToString());
            }
        }
        
        return stack.Pop();
    }
    
    static void Main() {
        string postfix = "ab+cd*-";
        
        Console.WriteLine("=== Postfix to Prefix ===\n");
        Console.WriteLine("Postfix Expression: " + postfix);
        
        string prefix = PostfixToPrefixConversion(postfix);
        
        Console.WriteLine("Prefix Expression: " + prefix + "\n");
        Console.WriteLine("Complexity: O(n)");
    }
}

File: DSA/Stack/Code/PostfixToPrefix.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

// Postfix to Prefix in C++
bool isOperator(char c) {
    return (c == '+' || c == '-' || c == '*' || c == '/');
}

string postfixToPrefix(string postfix) {
    stack<string> s;
    
    for (char c : postfix) {
        if (isOperator(c)) {
            string op2 = s.top(); s.pop();
            string op1 = s.top(); s.pop();
            
            string temp = "";
            temp += c;
            temp += op1;
            temp += op2;
            s.push(temp);
        } else {
            string temp = "";
            temp += c;
            s.push(temp);
        }
    }
    
    return s.top();
}

int main() {
    string postfix = "ab+cd*-";
    
    cout << "=== Postfix to Prefix ===\n\n";
    cout << "Postfix Expression: " << postfix << "\n";
    
    string prefix = postfixToPrefix(postfix);
    
    cout << "Prefix Expression: " << prefix << "\n\n";
    cout << "Complexity: O(n)\n";
    
    return 0;
}
